pkg/webhooks/validation/controller/controller.go

```diff
@@ -0,0 +1,530 @@
+			queue.Add(req)
+		},
+	}
+}
+
+// precompute GVK for known types.
+var (
+	configGVK     = kubeApiAdmission.SchemeGroupVersion.WithKind(reflect.TypeOf(kubeApiAdmission.ValidatingWebhookConfiguration{}).Name())
+	endpointGVK   = kubeApiCore.SchemeGroupVersion.WithKind(reflect.TypeOf(kubeApiCore.Endpoints{}).Name())
+	deploymentGVK = kubeApiApp.SchemeGroupVersion.WithKind(reflect.TypeOf(kubeApiApp.Deployment{}).Name())
+)
+
+func findClusterRoleOwnerRefs(client kubernetes.Interface, clusterRoleName string) []kubeApiMeta.OwnerReference {
+	clusterRole, err := client.RbacV1().ClusterRoles().Get(clusterRoleName, kubeApiMeta.GetOptions{})
+	if err != nil {
+		scope.Warnf("Could not find clusterrole: %s to set ownerRef. "+
+			"The webhook configuration must be deleted manually.",
+			clusterRoleName)
+		return nil
+	}
+
+	return []kubeApiMeta.OwnerReference{
+		*kubeApiMeta.NewControllerRef(
+			clusterRole,
+			kubeApiRbac.SchemeGroupVersion.WithKind("ClusterRole"),
+		),
+	}
+}
+
+func New(o Options) (*Controller, error) {
+	return newController(o, filewatcher.NewWatcher, ioutil.ReadFile, nil)
+}
+
+func newController(
+	o Options,
+	newFileWatcher filewatcher.NewFileWatcherFunc,
+	readFile readFileFunc,
+	reconcileDone func(),
+) (*Controller, error) {
+	caFileWatcher := newFileWatcher()
+	if err := caFileWatcher.Add(o.WebhookConfigPath); err != nil {
+		return nil, err
+	}
+	if err := caFileWatcher.Add(o.CAPath); err != nil {
+		return nil, err
+	}
+
+	c := &Controller{
+		o:             o,
+		queue:         workqueue.NewRateLimitingQueue(workqueue.DefaultItemBasedRateLimiter()),
+		caFileWatcher: caFileWatcher,
+		readFile:      readFile,
+		reconcileDone: reconcileDone,
+		ownerRefs:     findClusterRoleOwnerRefs(o.Client, o.ClusterRoleName),
+	}
+
+	c.sharedInformers = informers.NewSharedInformerFactoryWithOptions(o.Client, o.ResyncPeriod,
+		informers.WithNamespace(o.WatchedNamespace))
+
+	webhookInformer := c.sharedInformers.Admissionregistration().V1beta1().ValidatingWebhookConfigurations().Informer()
+	webhookInformer.AddEventHandler(makeHandler(c.queue, configGVK, o.WebhookConfigName))
+
+	endpointInformer := c.sharedInformers.Core().V1().Endpoints().Informer()
+	endpointInformer.AddEventHandler(makeHandler(c.queue, endpointGVK, o.ServiceName))
+
+	deploymentInformer := c.sharedInformers.Apps().V1().Deployments().Informer()
+	deploymentInformer.AddEventHandler(makeHandler(c.queue, deploymentGVK, o.GalleyDeploymentName))
+
+	return c, nil
+}
+
+func (c *Controller) Start(stop <-chan struct{}) {
+	go c.startFileWatcher(stop)
+	go c.sharedInformers.Start(stop)
+
+	for _, ready := range c.sharedInformers.WaitForCacheSync(stop) {
+		if !ready {
+			return
+		}
+	}
+
+	req := &reconcileRequest{"initial request to kickstart reconciliation"}
+	c.queue.Add(req)
+
+	go c.runWorker()
+}
+
+func (c *Controller) startFileWatcher(stop <-chan struct{}) {
+	for {
+		select {
+		case ev := <-c.caFileWatcher.Events(c.o.WebhookConfigPath):
```

Comment:
nit: a bit confusing its called caFileWatcher but its watching the webhook config?

Reply:
Agreed - renamed.